### DISCUSSION TITLE
自动保全派驻算法建议

### DISCUSSION THREAD

**Discussion by zhangweijian97** [MaaAssistantArknights/MaaAssistantArknights#1354] [Ideas]
首先，保全派驻主要是为了拿模组升级材料，所以只要能打到第三关就可以了，不必要全通。这样的话，感觉还是有可能做出来的。

算法分成两个部分，共通逻辑 和 特定配置。

共通逻辑意思是，不管是什么地图，这个逻辑都不变。比如，最典型的，进入保全派驻任务的一个个点击的逻辑，就是共通逻辑。还有检测手牌的操作，放置干员的操作，都是共通逻辑。但除了操作之外，还有放置干员的策略，这个也是共通逻辑。也就是说，对于如何攻略保全派驻的策略，应该有一套共通逻辑，在这基础之上，配合特定逻辑进行微调。具体的后面说。

特定配置意思是，对于每一个任务，需要专门进行特定的配置。可以简单分为地图特定配置和干员特定配置。比如，钢铁萝卜的地图特定配置就是地图坐标，摆放干员的地面位置1，地面位置2，高台位置1，高台位置2，等。类似现在的肉鸽，摆放干员的位置是固定的，干员则是随机应变的。然后是干员特定配置，这个就是指的是大核，或者说核心干员。比如钢铁萝卜就是澄闪。可以有多个核心干员，比如配置，高台核心1，澄闪，高台核心2，异客，高台核心3，灵芝，地面核心1，泥岩。核心干员的放置逻辑和非核心干员不同。

我尝试把自己模拟为脚本，来进行描述：

- （前面一路点击进入游戏，这部分不重要，略过）
- 检测到一局游戏开始
- 步骤：替换4个干员；行动：不进行替换，直接开始游戏（这个策略后续可以优化）；下一步行动：获取手牌信息
- 步骤：获取手牌信息；行动：点击暂停，依次点击干员头像，获取当前手牌信息；下一步行动：检查手牌情况
- 步骤：检查手牌情况，判断是否符合开局基本手牌，如果不符合，重开游戏，如何符合，继续。（注1: 开局基本手牌，见文末）；下一步行动：部署先锋
- 步骤：部署先锋；行动：当费用足够后，部署先锋；下一步行动：合成高台大核；
- 步骤：合成高台大核；行动：一旦费用足够，依次下5个合成素材，以及最后部署高台大核；下一步行动：过牌
- 步骤：过牌；行动：一旦费用足够，部署手牌中的辅助到第三个高台位，直到手牌无辅助。每次部署后，暂停，获取手牌信息。
- 步骤：抽卡；行动：一旦调用凭证足够，点击抽卡。抽卡之后，暂停，获取手牌信息；下一步行动：合成地面大核
- 步骤：合成地面大核；行动：一旦费用足够，依次下地面合成干员，以及最后部署地面大核。（注：如果没满5个，即使手牌有大核，也不下。
  - 下一步行动有两种情况
  - 第一种，未能合成地面大核，那么下一步行动就包含三个，过牌，抽卡，自己（也就是合成地面大核）
  - 第二种，成功合成地面大核，那么下一步行动就是过牌，抽卡，真正的下一步（也就是合成第二个高台大核）
- 步骤：合成第二个高台大核，情况和 合成地面大核类似
- 步骤：合成第二个地面大核，情况和 合成地面大核类似

注1: 开局基本手牌：包含一个挡二先锋，一个大核，以及满足合成大核的5个任意干员，比如，一个推进之王，一个2技能澄闪，5个任意狙击。
注2: 等完成2个高台大核两个地面大核之后，就挂机了。

**Comment by ABA2396** [comment]
https://user-images.githubusercontent.com/99072975/173285576-519a54e9-0dbd-47f6-a13d-ea9e8b023334.mp4

做了个大概的演示
